hail/python/hail/backend/service_backend.py

```diff
@@ -1,157 +1,370 @@
+            user_local_reference_cache_dir=user_local_reference_cache_dir,
+            remote_tmpdir=remote_tmpdir,
+        )
+
+    def __init__(self,
+                 billing_project: str,
+                 sync_fs: FS,
+                 async_fs: AsyncFS,
+                 bc: hb.BatchClient,
+                 disable_progress_bar: bool,
+                 batch_attributes: Dict[str, str],
+                 user_local_reference_cache_dir: Path,
+                 remote_tmpdir: str):
+        self.billing_project = billing_project
+        self._sync_fs = sync_fs
+        self._async_fs = async_fs
+        self.bc = bc
+        self.async_bc = self.bc._async_client
+        self.disable_progress_bar = disable_progress_bar
+        self.batch_attributes = batch_attributes
+        self.user_local_reference_cache_dir = user_local_reference_cache_dir
+        self.remote_tmpdir = remote_tmpdir
 
-        self.socket = ServiceSocket(deploy_config=deploy_config)
+    @property
+    def fs(self) -> FS:
+        return self._sync_fs
 
     @property
     def logger(self):
-        return self._logger
+        return log
 
     @property
-    def fs(self) -> GoogleCloudStorageFS:
-        if self._fs is None:
-            from hail.fs.google_fs import GoogleCloudStorageFS
-            self._fs = GoogleCloudStorageFS()
-        return self._fs
-
     def stop(self):
-        self.socket.close()
+        pass
 
-    def _render(self, ir):
+    def render(self, ir):
         r = CSERenderer()
         assert len(r.jirs) == 0
         return r(ir)
 
     def execute(self, ir, timed=False):
-        resp = self.socket.request('execute',
-                                   code=self._render(ir),
-                                   billing_project=self._billing_project,
-                                   bucket=self._bucket)
+        return async_to_blocking(self._async_execute(ir, timed=timed))
+
+    async def _async_execute(self, ir, timed=False):
+        result = await self._async_execute_untimed(ir)
+        if timed:
+            return result, dict()
+        return result
+
+    async def _rpc(self,
+                   name: str,
+                   inputs: Callable[[afs.ReadableStream, str], Awaitable[Tuple[str, dict]]]):
+        token = secret_alnum_string()
+        iodir = TemporaryDirectory(ensure_exists=False).name  # FIXME: actually cleanup
+        with TemporaryDirectory(ensure_exists=False) as _:
+            async with await self._async_fs.create(iodir + '/in') as infile:
+                await inputs(infile, token)
+
+            batch_attributes = self.batch_attributes
+            if 'name' not in batch_attributes:
+                batch_attributes = {**batch_attributes, 'name': name}
+            bb = self.async_bc.create_batch(token=token, attributes=batch_attributes)
+
+            j = bb.create_jvm_job([
+                'is.hail.backend.service.ServiceBackendSocketAPI2',
+                os.environ['HAIL_SHA'],
```

Comment:
Where do these environment variables come from when a regular user is using the ServiceBackend?

Reply:
A regular user needs to set these environment variables for the service backend to work properly. When we're ready to actually release this to users, we'll need to be able to fall back on a sha and JAR url which are read from a file generated when we generate the wheel. I haven't decided exactly how to do that yet, but it will be something similar to how hailctl currently works.